chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "revertTipsetsFromHeightStmt", err)
+	}
+	si.countMessagesStmt, err = si.db.Prepare(stmtCountMessages)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "countMessagesStmt", err)
+	}
+	si.minNonRevertedHeightStmt, err = si.db.Prepare(stmtMinNonRevertedHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "minNonRevertedHeightStmt", err)
+	}
+	si.tipsetExistsNotRevertedStmt, err = si.db.Prepare(stmtTipsetExistsNotReverted)
+
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "tipsetExistsNotRevertedStmt", err)
+	}
+	si.tipsetExistsNotRevertedStmt, err = si.db.Prepare(stmtTipsetExistsNotReverted)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "tipsetExistsNotRevertedStmt", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) IndexEthTxHash(ctx context.Context, txHash ethtypes.EthHash, msgCid cid.Cid) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		return si.indexEthTxHash(ctx, tx, txHash, msgCid)
+	})
+}
+
+func (si *SqliteIndexer) IndexSignedMessage(ctx context.Context, msg *types.SignedMessage) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		return si.indexSignedMessage(ctx, tx, msg)
+	})
+}
+
+func (si *SqliteIndexer) indexSignedMessage(ctx context.Context, tx *sql.Tx, msg *types.SignedMessage) error {
+	if msg.Signature.Type != crypto.SigTypeDelegated {
+		return nil
+	}
+
+	ethTx, err := ethtypes.EthTransactionFromSignedFilecoinMessage(msg)
+	if err != nil {
+		return xerrors.Errorf("error converting filecoin message to eth tx: %w", err)
+	}
+
+	txHash, err := ethTx.TxHash()
+	if err != nil {
+		return xerrors.Errorf("error hashing transaction: %w", err)
+	}
+
+	return si.indexEthTxHash(ctx, tx, txHash, msg.Cid())
+}
+
+func (si *SqliteIndexer) indexEthTxHash(ctx context.Context, tx *sql.Tx, txHash ethtypes.EthHash, msgCid cid.Cid) error {
+	insertEthTxHashStmt := tx.Stmt(si.insertEthTxHashStmt)
+	_, err := insertEthTxHashStmt.ExecContext(ctx, txHash.String(), msgCid.Bytes())
+	if err != nil {
+		return xerrors.Errorf("failed to index eth tx hash: %w", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) Apply(ctx context.Context, from, to *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	// We're moving the chain ahead from the `from` tipset to the `to` tipset
+	// Height(to) > Height(from)
+	err := withTx(ctx, si.db, func(tx *sql.Tx) error {
+		// index the `to` tipset first as we only need to index the tipsets and messages for it
+		if err := si.indexTipset(ctx, tx, to); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		// index the `from` tipset just in case it's not indexed
```

Comment:
is this really necessary?

when we do events we'll have to care about `from` but is there a good reason to do this here?

Reply:
Removed these in both `Apply` and `Revert`.